cmd/ncproxy/run.go

```diff
@@ -0,0 +1,178 @@
+	// ncproxy's commandline to launch with the -run-service flag set.
+	if *unregisterSvc {
+		if *registerSvc {
+			return errors.New("-register-service and -unregister-service cannot be used together")
+		}
+		return unregisterService()
+	}
+
+	if *registerSvc {
+		return registerService()
+	}
+
+	// Provider ID: cf9f01fe-87b3-568d-ecef-9f54b7c5ff70
+	// Hook isn't closed explicitly, as it will exist until process exit.
+	if hook, err := etwlogrus.NewHook("Microsoft.Virtualization.NCProxy"); err == nil {
+		logrus.AddHook(hook)
+	} else {
+		logrus.Error(err)
+	}
+
+	// Register our OpenCensus logrus exporter
+	trace.ApplyConfig(trace.Config{DefaultSampler: trace.AlwaysSample()})
+	trace.RegisterExporter(&oc.LogrusExporter{})
+
+	ctx := context.Background()
+	conf, err := loadConfig(*configPath)
+	if err != nil {
+		return errors.Wrap(err, "failed getting configuration file")
+	}
+
+	if conf.GRPCAddr == "" {
+		return errors.New("missing GRPC endpoint in config")
+	}
+
+	if conf.TTRPCAddr == "" {
+		return errors.New("missing TTRPC endpoint in config")
+	}
+
+	// If there's a node network service in the config, assign this to our global client.
+	if conf.NodeNetSvcAddr != "" {
+		log.G(ctx).Infof("Connecting to NodeNetworkService at address %s", conf.NodeNetSvcAddr)
+
+		dialCtx := ctx
+		opts := []grpc.DialOption{grpc.WithInsecure(), grpc.WithStatsHandler(&ocgrpc.ClientHandler{})}
+		if conf.Timeout > 0 {
+			var cancel context.CancelFunc
+			dialCtx, cancel = context.WithTimeout(ctx, time.Duration(conf.Timeout)*time.Second)
+			defer cancel()
+			opts = append(opts, grpc.WithBlock())
+		}
+		client, err := grpc.DialContext(dialCtx, conf.NodeNetSvcAddr, opts...)
+		if err != nil {
+			return fmt.Errorf("failed to connect to NodeNetworkService at address %s", conf.NodeNetSvcAddr)
+		}
+
+		log.G(ctx).Infof("Successfully connected to NodeNetworkService at address %s", conf.NodeNetSvcAddr)
+
+		netSvcClient := nodenetsvc.NewNodeNetworkServiceClient(client)
+		nodeNetSvcClient = &nodeNetSvcConn{
+			addr:     conf.NodeNetSvcAddr,
+			client:   netSvcClient,
+			grpcConn: client,
+		}
+	}
+
+	log.G(ctx).WithFields(logrus.Fields{
+		"TTRPCAddr":      conf.TTRPCAddr,
+		"NodeNetSvcAddr": conf.NodeNetSvcAddr,
+		"GRPCAddr":       conf.GRPCAddr,
+		"Timeout":        conf.Timeout,
+	}).Info("starting ncproxy")
+
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, os.Interrupt, syscall.SIGTERM)
+	defer signal.Stop(sigChan)
+
+	// Create new server and then register NetworkConfigProxyServices.
+	server, err := newServer(ctx, conf)
+	if err != nil {
+		return errors.New("failed to make new ncproxy server")
+	}
+
+	ttrpcListener, grpcListener, err := server.setup(ctx)
+	if err != nil {
+		return errors.New("failed to setup ncproxy server")
+	}
+
+	serveErr := make(chan error, 1)
+
+	// Launch as a Windows Service if necessary
+	if *runSvc {
+		// Setup panic log file
+		panicLog := filepath.Join(filepath.Dir(os.Args[0]), "ncproxy-panic.log")
```

Comment:
What is ncproxy's working directory?

Reply:
Guessing it's the bundle dir, which would make getting the log out difficult, and that's why we're using the exe dir here? Can we just take a log dir as part of ncproxy's config?

Reply:
Sure

Reply:
Done, right now the only thing that will go in the log dir is panic logs. We should add a flag to specify the type of logging we'd like to do (etw vs file etc.). This would help for local dev definitely